nav2_system_tests/src/system/test_system_launch.py

```diff
@@ -40,15 +42,24 @@ def generate_launch_description():
     bringup_dir = get_package_share_directory('nav2_bringup')
     params_file = os.path.join(bringup_dir, 'params', 'nav2_params.yaml')
 
-    # Replace the `use_astar` setting on the params file
-    param_substitutions = {
-        'planner_server.ros__parameters.GridBased.use_astar': os.getenv('ASTAR')}
+    # Replace the default parameter values for testing special features
+    # without having multiple params_files inside the nav2 stack
+    context = LaunchContext()
+    param_substitutions = {}
+
+    if (os.getenv('ASTAR') == 'True'):
+        param_substitutions.update({'use_astar': 'True'})
+
+    if (os.getenv('GROOT_MONITORING') == 'True'):
+        param_substitutions.update({'enable_groot_monitoring': 'True'})
+
     configured_params = RewrittenYaml(
         source_file=params_file,
         root_key='',
         param_rewrites=param_substitutions,
         convert_types=True)
 
+    new_yaml = configured_params.perform(context)
```

Comment:
I'm fine with this, I'm a little confused by no other launch file does this though. Are we misusing it elsewhere?

Reply:
I tried to figure it out yesterday. But I'm like 95% certain that is does not work elsewhere without the perform. (I remember beeing 100% certain but I investigated into this deeply a few weeks ago..). Especially nested stuff like `ros__parameters.nav2_bt_navigator.stuff.key` is certainly not working but living in some launch files.

I saw a few occasions of seemingly people tried to use it and then swapped to directly addressing it through `Declare_Argument` so RewrittenYaml is not being used. It sometimes even sits there and is not used at all..

Maybe a thing for a dedicated issue..?

In this occasion I remember that I printed the whole original yaml at Line 61 and it was not changed. Only after perform you get a filepath to a temporal file containing the content of original yaml with your intended alterations. (But the original yaml stays the same.. At least the one in src. I did not check `install/nav2_bringup/share/...`)


------------------

I just revisited the RewrittenYaml.py .. If it is not miraculously triggered from a complete different level, than I'm 100% confident that this thing does nothing without calling `perform`. Therefor, all other occasions of this without a perform (none as I last checked..??) should not have any effect..

I'm up for an alternate opinion on that.. We should probably introduce some tests that check if the launch files perform with the arguments we expect them to run with..?

Reply:
So then should we fix something here (not in this PR)? Or if its not being used in other launch files, remove it? I didn't write this script and I was never 100% clear why we needed it over other options so this is something I'm not certain about.

Reply:
Burning questions from @SteveMacenski on this that should be answered before taking it into another PR/Issue:

> is the rewritten thing in other launch files just useless and should be removed?

No, I do not think it is useless. Especially for tests it is great to have a way to use the default yaml with a parameters to "default" but want to alter some of them for your test cases (in this instance: enabling Groot or changing the default BT)
Could also be handy for adding/removing the currently loaded plugin composition of i.e. planners (but only altering the [array] of plugins and not adding a full section with dedicated parameters for each plugin.
In this instance I was also able to activate the A-Start tests that were locked/not used before.

Also altering `use_sim_time` could become pretty handy.. Think of it as having `using_sim_time` by default off (real robot) and programmatically switching it on with the bringup tb3_simulation launch file.

On the other side: You could add for everything a dedicated Launch Argument, but looking at the load of variation in the nav2_param.yaml I would strongly refrain from doing so. This easily blows up a launch file, at least in my opinion.

> Is it useful, but we didnt properly use it so worth fixing?

I highly think so in both cases. At the time of my investigations I was not aware of anybody using it with the `perform()` function and therefor concluding that it currently is just dead code.
It probably is not the smallest task but something highly recommended by me to list it in a coming up testing tutorial for the docs, as I see here the biggest usecase.
For reallife scenarios I always think that it's faster to have separated and tuned yamls run by some sort of autostart. Only one place should hold all your parameters.

Maybe in some corner cases where programmatic alteration of multiple repeating instances, this would also come in handy:

- Increase all speeds by factor X
- Update all my special API-keys for all your plugins

But those are also probably just things during tuning / testing.

> We don't know

Things that should be done as next step:

0. make a dedicated issue / PR for this
1. go through all instances of RewrittenYaml and look if it is used correctly by not only looking but also testing it
2. remove / fix the broken instances of it
3. add the feature in documentation
4. Use it in actual test scenarios
5. Improve it and/or try to merge it into the Launch system (there exists Substitutions, but this is also hard to grasp without any proper Documentation -> RewrittenYaml had some unique advantages over Substitutions, at least from the undocumented user perspective)